src/apps/external_executor/external_executor.cpp

```diff
@@ -707,37 +709,72 @@ namespace externalexecutor
       auto sub = [this](
                    ccf::endpoints::CommandEndpointContext& ctx,
                    temp::Event&& payload,
-                   ccf::grpc::StreamPtr<temp::EventInfo>&& out_stream) {
+                   ccf::grpc::StreamPtr<temp::SubResult>&& out_stream) {
         std::unique_lock<ccf::pal::Mutex> guard(subscribed_events_lock);
 
-        subscribed_events.emplace(std::make_pair(
-          payload.SerializeAsString(),
-          ccf::grpc::detach_stream(std::move(out_stream))));
-        LOG_INFO_FMT("Subscribed to event {}", payload.SerializeAsString());
+        auto it = subscribed_events.find(payload.name());
+        if (it != subscribed_events.end())
+        {
+          LOG_INFO_FMT("Returning sub error");
+          ccf::grpc::set_grpc_response_trailers(
+            ctx.rpc_ctx,
+            ccf::grpc::make_grpc_status(
+              GRPC_STATUS_FAILED_PRECONDITION,
+              fmt::format(
+                "Already have a subscriber for {} - only support a single "
+                "subscriber per-event",
+                payload.name())));
```

Comment:
I'm not keen on the fact that we're now leaking the HTTP/2 abstraction of gRPC (HTTP/2 trailers are used to pass gRPC status code/message to client) to the app. 

I understand that it was not possible to simply return `ccf::grpc::make_error();` from this handler, without having to specify `Empty` somewhere? Could the wrapper for streaming endpoint not return a different type, e.g. `GrpcAdapterNoResponse`?

Reply:
This is now done - the streaming return type is `GrpcAdapterStreamingResponse`, which can no longer be any kind of success value so doesn't need to be templated. Still need a trailing return type on endpoints which return both errors and `make_pending` in different paths, but that's fine.